### ISSUE TITLE
Array.prototype.filter definition forces a boolean

### ISSUE THREAD

**Issue by denis-sokolov** [Microsoft/TypeScript#5850]
It is natural to use `Array.prototype.filter` to filter items with missing information:

```typescript
type Foo = { prop: string }

const a: Foo[] = [
	{ prop: 'bar' },
	{ prop: null },
	{ prop: 'baz' }
] 

a.filter(x => x.prop)
```

Unfortunately, this [does not compile](http://www.typescriptlang.org/Playground#src=%0Atype%20Foo%20%3D%20%7B%20prop%3A%20string%20%7D%0A%0Aconst%20a%3A%20Foo%5B%5D%20%3D%20%5B%0A%09%7B%20prop%3A%20'bar'%20%7D%2C%0A%09%7B%20prop%3A%20null%20%7D%2C%0A%09%7B%20prop%3A%20'baz'%20%7D%0A%5D%20%0A%0Aa.filter(x%20%3D%3E%20x.prop)%0A), as `filter` definition requires the result of the filter function to be strictly boolean. Would it make sense to make the `filter` definition less strict?

**Comment by mhegazy** [comment]
how about:
```ts
a.filter(x => !! x.prop)
```

Alternatively, you can define a new signature for filter in a file in your project as such:

```ts
interface Array<T> {
    filter(callbackfn: (value: T, index: number, array: T[]) => any, thisArg?: any): T[];
}
```

**Comment by denis-sokolov** [comment]
> how about `a.filter(x => !! x.prop)`

Well, yes, type casting is the current way to appease the compiler.
My point was that we probably would want to not need to appease the compiler.

> you can define a new signature for filter in a file in your project

Sure, thanks for the suggestion. I know how to do that.

My point was not how to workaround this limitation in my project, I have no problem with that.
My point was to suggest an improvement to the core for the benefit of everybody.

**Comment by DanielRosenwasser** [comment]
We've run into this sort of problem with `lib.d.ts` a few times now (e.g. #4002). JavaScript is technically okay with a lot of things, but basically it comes down to trying to help users catch mistakes when we can. Forgetting to add a condition is an easy mistake, and `!!` is such an easy remedy for when you want to say "It's okay, I know what I'm doing."

If we did ever want to fix this, we couldn't just change the return type to `any`, because users could accidentally forget to return something in their callback. A more appropriate return type would be the (fairly awkward) `boolean | string | number | symbol | Object`.

**Comment by denis-sokolov** [comment]
@DanielRosenwasser, you’re right, this is a judgement call about inconvenience vs catching errors.
It is up to you, of course, to make the call.

Do consider, that in my point of view, `.filter` is equivalent to an `if` statement. And surely you [do not plan](http://www.typescriptlang.org/Playground#src=const%20foo%20%3D%20'value'%0Aif%20(foo)%20%7B%0A%09%0A%7D) to enforce mandatory type conversion in an if condition? But the same could be said about that: ‘Forgetting to perform some condition is an easy mistake (`if (foo) {}`), and `!!` is such an easy remedy (`if (!!foo) {}`).’

**Comment by adidahiya** [comment]
+1, it is inconsistent to have different levels of strictness for `if` and `filter`. I'd love to implement a TSLint rule to flag implicit coercion for users who want the additional safety though: https://github.com/palantir/tslint/issues/625

**Comment by RyanCavanaugh** [comment]
Approved. @mhegazy said there's no other suspicious use of `boolean` in lib.d.ts but someone should verify (`Array#some` ?).

Return type of the callback should be `any`

**Comment by DanielRosenwasser** [comment]
Why should the return type be `any` if you want to catch errors where users forget to return a type?

**Comment by RyanCavanaugh** [comment]
I was arguing for `{}`; @ahejlsberg argued that not return a value is a valid always-false filter (I hope that's not a total strawman argument?)

**Comment by RyanCavanaugh** [comment]
Oh, and `any` is consistent with this behavior in terms of non-returning functions:
```ts
function fn() { }
if (fn()) { // <-- not an error
}
```

**Comment by DanielRosenwasser** [comment]
Why would you ever do that instead of using `forEach`? Tons of things are valid in JavaScript, the point is to catch things that are clearly wrong.

**Comment by Elephant-Vessel** [comment]
I use typescript because I like static typing, I'm not sure I like this. If this is ok, who'd say that no more ugly type coersions in javascript won't diffuse into typescript "because we can do it in javascript"?

I'd counter-propose a breaking change for the if and require an explicit boolean there instead. In the end, typescript is all about adding the layer of typing that JS is missing right?.

**Comment by mjohnsonengr** [comment]
@Elephant-Vessel I'd agree with you were this some other kind of type coersion, but boolean coersion in the form of truthy and falsey values has become a staple of the language for many TypeScript developers, and doesn't detract from the purpose of the language at all.

Your breaking change proposal would break such a large portion of TypeScript codebases out there.  Would you also propose breaking changes for && and || so I can't do this:

    var propOrDefault = someThing && someThing.itsProp || defaultProp;

**Comment by Elephant-Vessel** [comment]
A coherent conceptual model of a language is more important for me than some neat shorthands - So yes I do, that should return either true or defaultProp,

Of course I was rhetorical about actually making this a breaking change, but it could at least be a compiler setting for us that don't want these compromises.

**Comment by mhegazy** [comment]
Fixed in #7779
